src/frame/decompress.rs

```diff
@@ -1 +1,312 @@
+use std::{fmt, hash::Hasher, io, mem::size_of};
+use twox_hash::XxHash32;
 
+use super::header::{BlockInfo, BlockMode, FrameInfo, MAX_FRAME_INFO_SIZE, MIN_FRAME_INFO_SIZE};
+use super::Error;
+use crate::block::WINDOW_SIZE;
+
+/// A reader for decompressing the LZ4 frame format
+///
+/// This Decoder wraps any other reader that implements `io::Read`.
+/// Bytes read will be decompressed according to the [LZ4 frame format](
+/// https://github.com/lz4/lz4/blob/dev/doc/lz4_Frame_format.md).
+///
+/// This reader can potentially make many small reads from the underlying
+/// stream depending on its format, therefore, passing in a buffered reader
+/// may be beneficial.
+pub struct FrameDecoder<R: io::Read> {
```

Comment:
Is there some advantage exposing `FrameDecoder` instead of having something like

``` pub fn decompress_into(input: &[u8], out: impl Writer) -> Result<(), Error> {```

Reply:
Yes, having `FrameDecoder` allow the user to wrap any `io::Read` and use that anywhere that takes an `io::Read`.

Reply:
I mean from an API it would be simpler to have
`pub fn decompress_into(input: impl Read, out: impl Write) -> Result<(), Error> {`

does `FrameDecoder` add some functionality we need or can we keep it private?

Reply:
It doesn't expose additional functionality but it's _existence_ is useful because it implements `io::Read`. For instance, you can use it on a function like https://docs.serde.rs/serde_json/fn.from_reader.html

Reply:
I'm thinking about the _public_ API. Locally I added this to frame/mod.rs, This works fine and has a simple API. Do we get additional functionality by exposing `FrameDecoder`?

```

/// Decompress all bytes of `input` into a new vec.
pub fn decompress_into(input: impl Read, mut out: impl Write) -> Result<(), Error> {
    let mut de = FrameDecoder::new(input);
    io::copy(&mut de, &mut out)?;
    Ok(())
}

/// Compress all bytes of `input`.
pub fn compress_with_into(
    frame_info: FrameInfo,
    mut input: impl Read,
    out: impl Write,
) -> Result<u64, std::io::Error> {
    let mut enc = FrameEncoder::with_frame_info(frame_info, out);
    let bytes_written = io::copy(&mut input, &mut enc)?;
    enc.finish()?; // todo missing last bytes
    Ok(bytes_written)
}

#[test]
fn test_frame() {
    use std::fs;
    let input = fs::File::open("./benches/compression_66k_JSON.txt").unwrap();
    let output = fs::File::create("./tmp.file").unwrap();
    compress_with_into(FrameInfo::default(), input, output).unwrap();
    let input = fs::File::open("./tmp.file").unwrap();
    let restore = fs::File::create("./tmp.restore.file").unwrap();
    decompress_into(input, restore).unwrap();
    let restore_data = fs::read("./tmp.restore.file").unwrap();
    let original = fs::read("./benches/compression_66k_JSON.txt").unwrap();
    assert_eq!(restore_data, original);
}
```

Reply:
Yes. Like in my example, you want use it as an io::Read directly.

I've used this many times and it was the motivator for me to write this in the first place.

```rust
fn test() {
    let compressed_input = std::fs::File::open("datafile").unwrap(); // or any other io::Read like a &[u8]
    let mut decompressed_input = lz4_flex::frame::FrameDecoder(compressed_input).unwrap();
    let json: serde_json::Value = serde_json::from_reader::<serde_json::Value>(decompressed_input).unwrap();

    // or if you have many json values in the input
    let compressed_input = std::fs::File::open("datafile").unwrap(); // or any other io::Read like a &[u8]
    let mut decompressed_input = lz4_flex::frame::FrameDecoder(compressed_input).unwrap();
    loop {
        match serde_json::from_reader::<serde_json::Value>(decompressed_input) {
            Ok(json) => { println!("json {:?}", json); }
            Err(e) if e.is_eof() => break,
            Err(e) => panic!("{}", e),
        }
        
    }
}
```

This is a read example, but the write side can use an FrameEncoder as well.

Reply:
Okay I get it, FrameDecoder is better to connect multiple streams. 

I think this is also a nice example to put into the docs

Reply:
Added examples to FrameEncoder/Decoder